005-http-bridge-metrics.md

```diff
@@ -0,0 +1,75 @@
+
+![Requests by HTTP method](/images/005-http-bridge-metrics/requests_by_method.png)
+
+The requests rate in total and even grouped by response codes (2XX, 4XX, 5XX).
+
+![Requests and Response codes](/images/005-http-bridge-metrics/requests_by_code.png)
+
+Bytes received /sec and Bytes sent /sec.
+
+![Bytes received](/images/005-http-bridge-metrics/bytes_received.png)
+
+![Bytes sent](/images/005-http-bridge-metrics/bytes_sent.png)
+
+Furthermore, I used an HTTP requests metrics with different regex for getting information on different endpoints.
+These HTTP requests/sec metrics bring labels like `code` (HTTP response code), `local` (bridge server listening), `remote` (client issuing the request), `method`, and `path`.
+I created each one based on `method` and `path` as the following (which reflect some of the operations we have on the bridge).
+
+* SEND to `/topics/<topic>`
+* SEND TO PARTITIONS to `/topics/<topic>/partitions/<partition>`
+* CREATE CONSUMER to `/consumers/<group>`
+* SUBSCRIBE to `/consumers/<group>/instances/<consumer>/subscription`
+* DELETE CONSUMER to `/consumers/<group>/instances/<consumer>`
+* POLL to `/consumers/<group>/instances/<consumer>/records`
+* COMMIT to `/consumers/<group>/instances/<consumer>/offsets`
+
+![HTTP requests](/images/005-http-bridge-metrics/http_requests.png)
+
+Maybe for the above, it could be useful just doing the sum and not having the rate for each `remote` client on each `path` and `method`.
+I was thinking about filtering them by topics but it’s possible only for the producer part because the topic is in the `path` while when a consumer subscribes to a topic, it’s specified in the body which is not available from this metrics; maybe it could be part of a custom metric to add.
+
+Maybe using another available metric about response time for processing each HTTP request (always provided with same label).
+
+They are not reported here but we are also getting JVM related metrics as for the operators about memory, GC, and so on.
+
+# Custom bridge metrics
+
+While the above metrics provides information about the HTTP part of the bridge, we are lacking the Kafka part.
+We could add some custom metrics which could have more info than the ones provided via HTTP requests:
+
+* Created consumers
+    * Could bring labels like “consumergroup”, “instanceid”, “base_uri” (not sure if this one is really useful). It could increase a consumers counter on the dashboard.
+* Subscription (requests)
+    * Could bring labels as “topics” but it should be a list, how to leverage that? What when the client uses “topic_pattern” and not specific topics in the request?
+* Deleted consumer
+    * Could bring labels like “consumergroup”, “instanceid”. It could decrease the previous consumers counter on the dashboard.
+* Poll (requests)
+    * Could bring labels like “consumergroup”, “instanceid”, number of “records” polled
+* Send (requests)
+    * Could bring labels like “topic” and number of “records” sent
```

Comment:
Do we have to add these? Can't we just expose the metrics from the Kafka client?

Reply:
That could be a way but then we need to expose them to Prometheus. The way we do this for Kafka, ZK etc is using JMX exporter which starts an HTTP server by itself. The bridge has its own HTTP server which already exposes the HTTP related metrics on the `/metrics` endpoint. It sounds to me that JMX exporter is not usable (if we don't want to have another running HTTP server listening on different port ... but then it means Prometheus scraping the bridge metrics on two HTTP servers ... sounds weird and not sure if possible). Maybe we could access the JMX beans programmatically and exposing them on the bridge HTTP server in a Prometheus format.

Reply:
I had a try using the `JmxCollector` library provided by the official Prometheus JMX exporter and I was able to get Kafka clients related metrics and then publishing them through the HTTP server `/metrics` endpoint avoiding the need for JMX exporter as an agent or as a standalone HTTP server. Without any configuration it exposes ALL the metrics but I guess that at this point it could be useful providing the bridge a YAML JMX config file to filter metrics. It would do in a `metrics` section in the `KafkaBridge` custom resource as well. @scholzj wdyt?

Reply:
So, does it provide the Kafka metrics together with the Vert.x based metrics?

Reply:
I get metrics from two different sources, the `PrometheusMeterRegistry` from Vert.x with the HTTP server ones and my JMX to Prometheus stuff for the Kafka side and I can chain them together.

Reply:
Ok ... so they are all in the same file / endpoint, right? i wonder how weird it would be if the `.spec.metrics` section in `KafkaBridge` configures only half of them. That is probably my only concern.

Reply:
That could be a concern but we can work in this way. The `.spec.metrics` is related to JMX stuff as it is for `Kafka.spec.kafka` while we could have something under `.spec.http.metrics` for HTTP related metrics configuration; anyway, right now there is nothing to be configurable there. The same could be in `.spec.amqp.metrics`.